### ISSUE TITLE
diag(p3): specialist BC-fit alone — can the policy network even represent the specialist?

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#272]
## Hypothesis being tested

Before debating WHY PPO can't find the specialist, we should verify that **the policy network architecture is even capable of representing the specialist's behavior**. If supervised behavioral cloning loss plateaus high, we have a representational gap — no RL algorithm will fix that.

This is a 1-hour sanity check that protects against an embarrassing failure mode.

## Protocol

1. Gather 10k+ specialist demonstrations on `minimal_specialization` (state → action).
2. Train the standard PPO policy network architecture (post-#216 `[T, N, obs_dim]` shape, 3-head action) to predict specialist actions via supervised cross-entropy loss.
3. Track BC loss and BC eval accuracy across epochs.

## Verdict logic

| BC outcome | Interpretation |
|---|---|
| **BC loss → 0, eval accuracy > 90%** | Architecture can represent specialist. PPO failure is in path-finding, not capacity. |
| **BC loss plateaus high, eval accuracy < 50%** | Architecture mismatch. Either hidden-size too small, missing nonlinearity, or fundamental capacity gap. |
| **BC converges to specialist behavior except in edge cases** | Architecture is mostly OK but lacks some inductive bias the specialist uses. Worth examining which states are misclassified. |

## Cost

~30 min including data gathering + training. No env rollouts during training — pure supervised.

## Why this matters

Cheap insurance. If we discover BC can't fit the specialist, all our PPO experiments are red herrings — fix the architecture first.

## Acceptance criteria

- [ ] Demonstration generator producing 10k+ (obs, specialist_action) pairs
- [ ] BC training script in `experiments/p3_specialization/bc_fit_only.py`
- [ ] Reports final BC loss + eval accuracy
- [ ] Notebook entry with verdict
- [ ] If verdict shows mismatch: file architecture-fix issue

## References

- Sister issue #270 (BC-init then PPO)
- Sister issue #271 (random-init MLP best-of-N)
- docs/ppo-help-brief.md

**Comment by rjwalters** [comment]
## Curator Enhancement

The issue's hypothesis and verdict table are already crisp; this comment fills in concrete file paths, hyperparameters, and architecture math so a Builder can start implementation without further exploration.

### Architecture math (post-#216 / post-#236)

The PPO actor is `bucket_brigade.training.networks.PolicyNetwork`:
- `obs_dim = 46` for `num_agents=4` (per `joint_trainer.py` docstring lines 23-32: 10 houses + 4 signals + 4 locations + 8 last_actions + 10 scenario_info + 4 identity one-hot)
- `action_dims = [10, 2, 2]` — three independent categorical heads (house, mode, signal), 14 logits total
- `hidden_size = 64` (default), shared 2-layer ReLU trunk -> per-head linear

**Parameter count**: `46x64 + 64x64 + 3 heads (64x10 + 64x2 + 64x2) + value(64x1)` = ~8.1k params. Plenty of capacity to memorize a deterministic function of one input slice (`obs["houses"]`) — but **the specialist requires per-agent ownership conditioning**, which only enters through the 4-bit identity one-hot tail. Smallest plausible mismatch the BC fit might reveal: the trunk fails to combine identity bits with house-state bits cleanly enough to route attention to owned houses. If so, this surfaces as: low loss on `mode`/`signal` heads (both trivial 2-way) but high loss on the `house` head (10-way, requires identity-conditioned argmin over burning-owned).

### Specialist policy (deterministic, recap)

From `bucket_brigade/baselines/specialist.py`:
```
if any owned house is BURNING -> [lowest_burning_owned_idx, WORK=1, signal=WORK]
else                          -> [lowest_owned_idx,        REST=0, signal=REST]
```
Ownership is round-robin: `owners = np.arange(10) % 4`. The function is a pure function of `obs["houses"]` and `agent_id`. Perfect BC fit is theoretically achievable since the mapping is deterministic and the agent identity is in the obs.

### Implementation sketch

**New file**: `experiments/p3_specialization/bc_fit_only.py`

1. **Demonstration generator** (~5 min CPU):
   - Construct env via `BucketBrigadeEnv` with `minimal_specialization` scenario, `num_agents=4`.
   - Roll random actions (or specialist actions — either gives diverse coverage; prefer specialist + epsilon-greedy noise so we hit some burning-owned states).
   - For each step, for each of 4 agents, record `(flatten_dict_obs(obs, agent_id=i, num_agents=4), specialist_action(obs, i, 4, 10))`.
   - Target ~10k pairs total (so ~2500 env steps).

2. **Supervised training**:
   - Instantiate `PolicyNetwork(obs_dim=46, action_dims=[10, 2, 2], hidden_size=64)`.
   - Loss: sum of `F.cross_entropy(logits[k], action[:, k])` across k in {0,1,2}.
   - Optimizer: Adam, `lr=3e-4`, `batch_size=64`.
   - Epochs: 10 (with ~125 batches/epoch on 8k train set, ~1250 SGD steps — plenty for an 8k-param net).
   - Train/eval split: 80/20 random, fixed seed.

3. **Reporting**:
   - Per-epoch: train loss, eval loss, eval per-head accuracy (3 numbers), eval joint accuracy (all-3-match).
   - Final printout in a format that can be pasted into a notebook entry.
   - Optional: confusion matrix on the `house` head segmented by `mode == WORK` (the only non-trivial case).

### Hyperparameters (defaults; override via CLI flags)

| Param | Value | Why |
|---|---|---|
| `lr` | `3e-4` | Adam default for small MLPs; stable for 8k-param model |
| `batch_size` | `64` | Matches PPO minibatch; small enough for CPU |
| `epochs` | `10` | Loss should plateau well before this on a deterministic mapping |
| `n_demonstrations` | `10000` | ~2500 env steps x 4 agents |
| `train_frac` | `0.8` | Standard |
| `seed` | `0` | Reproducibility |
| `device` | `"cpu"` | No rollouts during training; CPU is faster end-to-end than CPU<->GPU shuffling for this size |

### Expected runtime

~5 min data gen + ~10 min training on CPU = ~15-30 min total. No env rollouts during training (pure supervised), so this is dramatically cheaper than the sister #270 (BC-then-PPO, ~3h) and ~equal to #271 (random search, ~30 min).

### Verdict thresholds (operationalized)

| Condition | Verdict label | Notebook action |
|---|---|---|
| Eval loss < 0.05 AND joint accuracy > 0.90 | **Representable** | Move on — PPO failure is path-finding, not capacity. Strengthens case for MAPPO/CTDE intervention. |
| Eval loss > 0.5 AND any head accuracy < 0.50 | **Architecture mismatch** | File follow-up issue: try `hidden_size=128`, deeper trunk, or `TransformerPolicyNetwork`. |
| Loss in between, accuracy 0.50-0.90 | **Inductive-bias gap** | Examine misclassified states — likely transitions where identity-conditioned routing fails. |

## Affected Files

- `experiments/p3_specialization/bc_fit_only.py` — new BC training script (entire implementation lives here)
- `bucket_brigade/baselines/specialist.py` — read-only; source of demonstration labels
- `bucket_brigade/training/networks.py` — read-only; `PolicyNetwork` instantiation target
- `bucket_brigade/training/joint_trainer.py` — read-only; canonical `flatten_dict_obs` helper for obs construction
- `bucket_brigade/envs/scenarios_generated.py` — read-only; `minimal_specialization_scenario` factory

## Test Plan

- [ ] Manual verification: Run `uv run python experiments/p3_specialization/bc_fit_only.py --epochs 10 --seed 0` and confirm final eval loss + per-head accuracies are printed.
- [ ] Sanity check: Verify the demonstration generator produces at least 10k pairs and that ~25-40% of label rows have `mode=WORK` (i.e., we actually see burning-owned states; if essentially all `REST`, increase `ignition_rate` exploration or pre-ignite for data gen).
- [ ] Reproducibility: Run twice with the same seed; final losses must match to ~3 decimals.
- [ ] Verdict logged: Update `experiments/p3_specialization/` notebook/markdown with the verdict per the threshold table above.
- [ ] Edge case: If verdict is **mismatch** (loss > 0.5), file an architecture-fix issue per the issue's existing acceptance criterion.
- [ ] Automated tests: N/A (research diagnostic, not production code).

## References

- Sister issue #270 (BC-init then PPO — basin-trap vs anti-attractor)
- Sister issue #271 (random-init MLP best-of-N vs trained PPO)
- PR #257 (random-init PPO baseline; current 15%-of-specialist plateau)
- `docs/ppo-help-brief.md` — broader Phase-1 trio motivation
- Project memory: PPO failure mode (env-side fixes verified in #197/#198/#199; independent-PPO still plateaus at 15% of optimum on `minimal_specialization`)
